### PR TITLE
[eclipse/xtext#1249] use bom for version management

### PR DESCRIPTION
**Pull request by szarnekow**
upstreamBranch must be reset.

### CODE REVIEW

File: org.eclipse.xtext.maven.plugin/src/test/resources/it/generate/bug463946/pom.xml [hunk 1]
```diff
@@ -115,6 +115,26 @@
...
+					</dependency>
+					<dependency>
+						<groupId>org.eclipse.emf</groupId>
+						<artifactId>org.eclipse.emf.codegen.ecore</artifactId>
+						<version>2.15.0</version>
+					</dependency>
+					<dependency>
+						<groupId>org.eclipse.emf</groupId>
+						<artifactId>org.eclipse.emf.codegen</artifactId>
+						<version>2.14.0</version>
```
Comment (cdietrich) [inline org.eclipse.xtext.maven.plugin/src/test/resources/it/generate/bug463946/pom.xml]: am not sure about that change

File: org.eclipse.xtext.maven.plugin/src/test/resources/it/generate/pom.xml [hunk 2]
```diff
@@ -7,9 +7,20 @@
...
-		<upstreamBranch>master</upstreamBranch>
+		<upstreamBranch>bom_issue1249</upstreamBranch>
 		<JENKINS_URL>http://services.typefox.io/open-source/jenkins</JENKINS_URL>
 	</properties>
+	<dependencyManagement>
+		<dependencies>
+			<dependency>
+				<groupId>org.eclipse.xtext</groupId>
+				<artifactId>xtext-dev.bom</artifactId>
+				<version>2.17.0-SNAPSHOT</version>
```
Comment (cdietrich) [inline org.eclipse.xtext.maven.plugin/src/test/resources/it/generate/pom.xml]: wonder how these tests work. what do they pick upstream xtext from. maybe a property for the version?

File: org.eclipse.xtext.maven.plugin/src/test/resources/it/generate/xcore-lang/pom.xml [hunk 3]
```diff
@@ -116,6 +116,26 @@
...
+						<version>2.15.0</version>
+					</dependency>
+					<dependency>
+						<groupId>org.eclipse.emf</groupId>
+						<artifactId>org.eclipse.emf.codegen.ecore</artifactId>
+						<version>2.15.0</version>
+					</dependency>
+					<dependency>
+						<groupId>org.eclipse.emf</groupId>
+						<artifactId>org.eclipse.emf.codegen</artifactId>
```
Comment (cdietrich) [inline org.eclipse.xtext.maven.plugin/src/test/resources/it/generate/xcore-lang/pom.xml]: see above

File: org.eclipse.xtext.maven.plugin/src/test/resources/it/generate/bug463946/pom.xml (hunk 1 above)
Comment (szarnekow) [reply org.eclipse.xtext.maven.plugin/src/test/resources/it/generate/bug463946/pom.xml]: The alternative is to add the bom to the generated pom, isn't it?

File: org.eclipse.xtext.maven.plugin/pom.xml [hunk 4]
```diff
@@ -61,13 +61,11 @@
 		<dependency>
 			<groupId>org.eclipse.emf</groupId>
 			<artifactId>org.eclipse.emf.ecore.xmi</artifactId>
-			<version>${emf.version}</version>
 			<scope>runtime</scope>
```
Comment (kthoms) [inline org.eclipse.xtext.maven.plugin/pom.xml]: could by managed in BOM

File: org.eclipse.xtext.maven.plugin/pom.xml [hunk 5]
```diff
@@ -61,13 +61,11 @@
...
 		</dependency>
 		<dependency>
 			<groupId>log4j</groupId>
 			<artifactId>log4j</artifactId>
-			<version>${log4j.version}</version>
 			<exclusions>
```
Comment (kthoms) [inline org.eclipse.xtext.maven.plugin/pom.xml]: could by managed in BOM

File: org.eclipse.xtext.maven.plugin/src/test/resources/it/generate/bug463946/pom.xml (hunk 1 above)
Comment (kthoms) [reply org.eclipse.xtext.maven.plugin/src/test/resources/it/generate/bug463946/pom.xml]: that's why we have the BOM. I'd prefer to use it.

**Comment by cdietrich** [comment]
there are now test failures

File: org.eclipse.xtext.maven.plugin/pom.xml (hunk 4 above)
Comment (cdietrich) [reply org.eclipse.xtext.maven.plugin/pom.xml]: sure? see test failures?

**Comment by cdietrich** [comment]
```
[DEBUG] -d /Users/dietrich/xtext-dev/git/xtext-maven/org.eclipse.xtext.maven.plugin/src/test/resources/it/generate/xcore-mapping/sample.emf/target/classes -classpath /Users/dietrich/xtext-dev/git/xtext-maven/org.eclipse.xtext.maven.plugin/src/test/resources/it/generate/xcore-mapping/sample.emf/target/classes:/Users/dietrich/xtext-dev/git/xtext-maven/.m2/org/eclipse/emf/org.eclipse.emf.ecore.xcore.lib/1.1.100/org.eclipse.emf.ecore.xcore.lib-1.1.100.jar:/Users/dietrich/xtext-dev/git/xtext-maven/.m2/org/eclipse/emf/org.eclipse.emf.ecore/2.12.0/org.eclipse.emf.ecore-2.12.0.jar:/Users/dietrich/xtext-dev/git/xtext-maven/.m2/org/eclipse/emf/org.eclipse.emf.common/2.12.0/org.eclipse.emf.common-2.12.0.jar: -sourcepath /Users/dietrich/xtext-dev/git/xtext-maven/org.eclipse.xtext.maven.plugin/src/test/resources/it/generate/xcore-mapping/sample.emf/src:/Users/dietrich/xtext-dev/git/xtext-maven/org.eclipse.xtext.maven.plugin/src/test/resources/it/generate/xcore-mapping/sample.emf/xcore-gen: -s /Users/dietrich/xtext-dev/git/xtext-maven/org.eclipse.xtext.maven.plugin/src/test/resources/it/generate/xcore-mapping/sample.emf/target/generated-sources/annotations -g -nowarn -target 1.8 -source 1.8 -encoding UTF-8
``` 

no xmi here

**Comment by cdietrich** [comment]
why this: https://github.com/eclipse/xtext-lib/blob/a84c150ad2b47ec10dfc10c5e50cb60d159ce99e/releng/org.eclipse.xtext.dev-bom/pom.xml#L196

File: org.eclipse.xtext.maven.plugin/src/test/resources/it/generate/pom.xml [hunk 6]
```diff
@@ -7,9 +7,20 @@
 	<artifactId>xtext-maven-plugin-it</artifactId>
 	<packaging>pom</packaging>
 	<properties>
-		<upstreamBranch>master</upstreamBranch>
+		<upstreamBranch>bom_issue1249</upstreamBranch>
```
Comment (cdietrich) [inline org.eclipse.xtext.maven.plugin/src/test/resources/it/generate/pom.xml]: am not sure if this is a good idea.
this is a client project.
am not sure if bom hurts here

Comment (cdietrich) [reply org.eclipse.xtext.maven.plugin/src/test/resources/it/generate/pom.xml]: but i am not a bom wizard

Comment (szarnekow) [reply org.eclipse.xtext.maven.plugin/src/test/resources/it/generate/pom.xml]: I recall that we tested this without the bom reference in the client projects and there were issues with dependencies if we do not import the bom on the "top-level".

File: org.eclipse.xtext.maven.plugin/src/test/resources/it/generate/bug463946/pom.xml (hunk 1 above)
Comment (szarnekow) [reply org.eclipse.xtext.maven.plugin/src/test/resources/it/generate/bug463946/pom.xml]: Can we do this in a second step please.

File: org.eclipse.xtext.maven.plugin/src/test/resources/it/generate/pom.xml (hunk 2 above)
Comment (szarnekow) [reply org.eclipse.xtext.maven.plugin/src/test/resources/it/generate/pom.xml]: I cannot answer that.

File: org.eclipse.xtext.maven.plugin/pom.xml (hunk 4 above)
Comment (kthoms) [reply org.eclipse.xtext.maven.plugin/pom.xml]: I'll check. I had visited the effective pom for xtend-maven-plugin with managed scope and looked fine to me. But could be that this was a wrong observation. When scopes and exclusions are not manageable (but why they shouldn't?), then I'll revert that in the bom.

File: org.eclipse.xtext.maven.plugin/src/test/resources/it/generate/bug463946/pom.xml (hunk 1 above)
Comment (kthoms) [reply org.eclipse.xtext.maven.plugin/src/test/resources/it/generate/bug463946/pom.xml]: For now it is OK to have fixed versions here. And maybe it is even better to pin versions here for a fixed test setup that does not change when the BOM changes. Same for several other test projects here. There is room for improvement, but this is not critical yet.

**Comment by szarnekow** [comment]
Ready to be merged.

**Pull request merged**